lib/convection/model/template.rb

```diff
@@ -183,9 +197,122 @@ def diff(other, stack_ = nil)
       end
 
       def to_json(stack_ = nil, pretty = false)
+        rendered_stack = render(stack_)
+        validate(rendered_stack)
+
         return JSON.generate(render(stack_)) unless pretty
         JSON.pretty_generate(render(stack_))
       end
+
+      def validate(rendered_stack = nil)
+        validate_resources(rendered_stack)
+        validate_mappings(rendered_stack)
+        validate_parameters(rendered_stack)
+        validate_outputs(rendered_stack)
+        validate_description(rendered_stack)
+        validate_bytesize(rendered_stack)
+
+      end
+
+      def validate_resources(rendered_stack = nil)
+        resources=rendered_stack["Resources"]
+        number_of_resources = resources.count
+        if number_of_resources > CF_MAX_RESOURCES
+          raise ArgumentError, "Error: Excessive Number of Resources (#{number_of_resources}) Max= #{CF_MAX_RESOURCES}"
+        end
+        largest_resource_name = resources.keys.max
+        if largest_resource_name == nil
+          largest_resource_name = ""
+        end
+        resource_name_characters = largest_resource_name.length
+        if  resource_name_characters> CF_MAX_RESOURCE_NAME
+          raise ArgumentError, "Error: Resource Name #{largest_resource_name} has too many characters (#{resource_name_characters}) Max=#{CF_MAX_RESOURCE_NAME} "
+        end
+      end
+
+      def validate_mappings(rendered_stack = nil)
+        mappings = rendered_stack ['Mappings']
+        number_of_mappings = mappings.count
+        if number_of_mappings > CF_MAX_MAPPINGS
+          raise ArgumentError, "Error: Excessive Number of Mappings (#{number_of_mappings}) Max= #{CF_MAX_MAPPINGS}"
+        end
+        mappings.each do |key,value|
+          if value.count > CF_MAX_MAPPING_ATTRIBUTES
+            raise ArgumentError, "Error: Excessive Number of Mapping Attributes"
+          end
+        end
+        mappings.keys.each do |key|
+          if key.length > CF_MAX_MAPPING_NAME
+            raise ArgumentError, "Error: Excessive Mapping Name #{key}"
+          end
+        end
+        mapping_attributes = mappings.values.flat_map do |inner_hash|
+          inner_hash.keys.select do |key|
+            value = inner_hash[key]
+          end
+        end
+        mapping_attributes.each do |attribute|
+          if attribute.length > CF_MAX_MAPPING_ATTRIBUTE_NAME
+            raise ArgumentError, "Error: Excessive Mapping Attribute Name #{attribute}"
+          end
+        end
+      end
+
+      def validate_parameters(rendered_stack = nil)
+        parameters= rendered_stack['Parameters']
+        number_of_parameters = parameters.count
+        if number_of_parameters > CF_MAX_PARAMETERS
+          raise ArgumentError, "Error: Excessive Number of Parameters (#{number_of_parameters}) Max= #{CF_MAX_PARAMETERS}"
+        end
+        largest_parameter_name = parameters.keys.max
+        if largest_parameter_name == nil
+          largest_parameter_name = ""
```

Comment:
See @simonirwin-r7's previous comment about using `||=` or use `.to_s` on max in which case nil will cause the result to be empty string.